openverse_catalog/dags/providers/provider_api_scripts/europeana.py

```diff
@@ -8,266 +8,172 @@
 
 Notes:                  https://www.europeana.eu/api/v2/search.json
 """
-
 import argparse
 import logging
 from datetime import datetime, timedelta, timezone
 
+import common
 from airflow.models import Variable
 from common.licenses import get_license_info
 from common.loader import provider_details as prov
-from common.requester import DelayedRequester
-from common.storage.image import ImageStore
-from requests.exceptions import JSONDecodeError
+from providers.provider_api_scripts.provider_data_ingester import ProviderDataIngester
 
 
-logging.basicConfig(
-    format="%(asctime)s - %(name)s - %(levelname)s:  %(message)s", level=logging.INFO
-)
 logger = logging.getLogger(__name__)
+logging.getLogger(common.urls.__name__).setLevel(logging.WARNING)
+
+
+class EuropeanaRecordBuilder:
+    """
+    A small class to contain the record building functionality
+    and simplify testing a bit.
+    """
+
+    def get_record_data(self, data: dict) -> dict:
+        record = {
+            "foreign_landing_url": self.get_foreign_landing_url(data),
+            "image_url": data.get("edmIsShownBy")[0],
```

Comment:
Is it possible for this (or `foreign_landing_url`, `foreign_id`, or `license_info`) to be `None`? We typically get the required fields first and return `None` early from `get_record_data` if they cannot be extracted.

Reply:
Are those fields allows to be None? :open_mouth:

Reply:
In other scripts, if a required value is None, we stop the current item's step early by using either `return None` or `continue`/`break` inside a loop. Here is an example:
https://github.com/WordPress/openverse-catalog/blob/1506d6d9d391d7d12c8f5c515737623218036928/openverse_catalog/dags/providers/provider_api_scripts/cleveland_museum.py#L46-L48

Reply:
I'd love a list of the required fields, maybe on the base class, to reference at the _end_ of the `get_record_data` method... or potentially as something that the base class itself checks. These kinds of checks seem like boilerplate and distributed, undocumented knowledge that the base class would be perfect for.

Reply:
The list of required fields is in the `ImageStore.add_item` method's docs (the same is available in the AudioStore):

https://github.com/WordPress/openverse-catalog/blob/1506d6d9d391d7d12c8f5c515737623218036928/openverse_catalog/dags/common/storage/image.py#L67-L77

Do you think there's a better place for this list?

The checks are boilerplate, but not boilerplate _enough_ to extract them because all API responses are different, and the required pieces of information can be located in different stages of the script. The idea behind returning `None` is to fail early and skip the unnecessary work of extracting the metadata (which can be quite tedious for some API responses) if the item does not have the required fields and will be discarded anyways.

Reply:
Staci also recently added documentation around this exact thing as part of #790, which can be found here: https://github.com/WordPress/openverse-catalog/blob/main/openverse_catalog/docs/data_models.md

It's also a goal of the Data Normalization project (I believe) to improve on this understanding and ensure that we have documentation around it.

Reply:
I'll also say here that these fields being None shouldn't cause errors or anything -- I believe the MediaStore should successfully discard them (if I'm wrong and that's not the case I agree we should add this behavior in the Data normalization project). The point of the None checks is mostly to halt processing of a record _early_ if it's just going to be discarded anyway, as a general best practice.

We could update the base class to, instead of having just an abstract `get_record_data`, have abstract `get_required_record_data` and `get_optional_record_data` methods. Or even individual abstract methods for fetching each required field (`get_license_info`, `get_foreign_identifier`...). Then we could make `get_record_data` on the base class do the None checks for us.

I'll make an issue.

Reply:
#834